formatter-gitlab/markdownlint-cli2-formatter-gitlab.js

```diff
@@ -0,0 +1,76 @@
+// @ts-check
+
+"use strict";
+
+const fs = require("node:fs").promises;
+const path = require("node:path");
+
+const { createHash } = require("node:crypto");
+
+/**
+ * @param {string} violation The complete textual description of the violation
+ * @returns {string} The SHA256 fingerprint for the violation
+ */
+const createFingerprint = function createFingerprint (violation) {
+  const sha256 = createHash("sha256");
+  sha256.update(violation);
+  return sha256.digest("hex");
+};
+
+// Writes markdownlint-cli2 results to a
+// GitLab code quality report artifact JSON file
+// eslint-disable-next-line max-len
+// see: https://docs.gitlab.com/ee/ci/testing/code_quality.html#implementing-a-custom-tool
+const outputFormatter = (options, params) => {
+  const { directory, results } = options;
+  const { name, spaces } = (params || {});
+
+  const issues = [];
+
+  for (const errorInfo of results) {
+    const { fileName, lineNumber, ruleNames, ruleDescription, errorDetail,
+      errorContext, errorRange } = errorInfo;
+
+    const ruleName = ruleNames.join("/");
+    const errorDetailText = errorDetail ? ` [${errorDetail}]` : "";
+    const text = `${ruleName}: ${ruleDescription}${errorDetailText}`;
+    const column = (errorRange && errorRange[0]) || 0;
+    const columnText = column ? `:${column}` : "";
+    const description = ruleDescription +
+          (errorDetail ? ` [${errorDetail}]` : "") +
+          (errorContext ? ` [Context: "${errorContext}"]` : "");
+    // construct error text with all details to use for unique fingerprint
```

Comment:
Is this your technique, or is this part of the specifications? Also, I think comments should begin with a capital letter.

Reply:
This is to create a unique fingerprint for the violation. I reused the code from the default formatter to produce the text.

The [code climate specification](https://github.com/codeclimate/platform/blob/master/spec/analyzers/SPEC.md#data-types) states for the `fingerprint`:

> A unique, deterministic identifier for the specific issue being reported to allow a user to exclude it from future analyses.